File: agent/app/domain/manifest.py
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class SecurityConfiguration(StrictModel):
    encryption_enabled: bool


class BackupConfiguration(StrictModel):
    enabled: bool


type PolicyConstraintName = Literal[
    "allowed_regions",
    "max_budget",
    "required_tags",
    "prohibited_resource_types",
    "required_encryption",
    "required_backup",
]


class PolicyApplication(StrictModel):
    requirements_loaded: Literal[True]
    applied_constraints: tuple[PolicyConstraintName, ...]


class Ec2Resource(StrictModel):
    type: Literal["aws_ec2"]
    name: str = Field(min_length=1, max_length=64, pattern=r"^[a-zA-Z0-9_-]+$")
    instance_type: str = Field(min_length=1, max_length=64)
    image: str = Field(min_length=1, max_length=128)
    count: int = Field(default=1, ge=1, le=20)


class RdsResource(StrictModel):
    type: Literal["aws_rds"]
    name: str = Field(min_length=1, max_length=64, pattern=r"^[a-zA-Z0-9_-]+$")
    engine: Literal["postgres", "mysql"]
    instance_class: str = Field(min_length=1, max_length=64)
    allocated_storage_gb: int = Field(ge=20, le=16384)


class S3Resource(StrictModel):
    type: Literal["aws_s3"]
    name: str = Field(min_length=3, max_length=63, pattern=r"^[a-z0-9][a-z0-9.-]+[a-z0-9]$")
    versioning: bool = True


AwsResource = Annotated[Ec2Resource | RdsResource | S3Resource, Field(discriminator="type")]


class ResourceManifest(StrictModel):
    schema_version: Literal["1.0"] = "1.0"
    provider: Literal["aws"] = "aws"
    region: str = Field(min_length=3, max_length=32)
    environment: Literal["development", "staging", "production", "sandbox"]
    monthly_budget_usd: float | None = Field(default=None, gt=0)
    tags: dict[str, str] = Field(default_factory=dict)
    security: SecurityConfiguration
    backup: BackupConfiguration
    policy: PolicyApplication
    resources: list[AwsResource] = Field(min_length=1, max_length=50)

File: packages/shared-contracts/src/schemas/manifest.ts
import { z } from "zod";

const identifier = z.string().min(1).max(64).regex(/^[a-zA-Z0-9_-]+$/);
const region = z.string().min(3).max(32).regex(/^[a-z]{2}-[a-z]+-\d$/);

const ec2 = z.object({
  type: z.literal("aws_ec2"),
  name: identifier,
  instance_type: z.string().min(1).max(64),
  image: z.string().min(1).max(128),
  count: z.number().int().min(1).max(20).default(1),
}).strict();

const rds = z.object({
  type: z.literal("aws_rds"),
  name: identifier,
  engine: z.enum(["postgres", "mysql"]),
  instance_class: z.string().min(1).max(64),
  allocated_storage_gb: z.number().int().min(20).max(16384),
}).strict();

const s3 = z.object({
  type: z.literal("aws_s3"),
  name: z.string().min(3).max(63).regex(/^[a-z0-9][a-z0-9.-]+[a-z0-9]$/),
  versioning: z.boolean().default(true),
}).strict();

/** Cross-layer source of truth for agent-generated AWS manifest candidates. */
export const ResourceManifestSchema = z.object({
  schema_version: z.literal("1.0").default("1.0"),
  provider: z.literal("aws").default("aws"),
  region,
  environment: z.enum(["development", "staging", "production", "sandbox"]),
  monthly_budget_usd: z.number().positive().optional(),
  tags: z.record(z.string()).default({}),
  resources: z.array(z.discriminatedUnion("type", [ec2, rds, s3])).min(1).max(50),
}).strict();

export type CanonicalResourceManifest = z.infer<typeof ResourceManifestSchema>;

export function validateResourceManifest(candidate: unknown):
  | { ok: true; manifest: CanonicalResourceManifest }
  | { ok: false; issues: Array<{ path: string; message: string }> } {
  const parsed = ResourceManifestSchema.safeParse(candidate);
  if (parsed.success) return { ok: true, manifest: parsed.data };
  return {
    ok: false,
    issues: parsed.error.issues.map((issue) => ({ path: issue.path.join("."), message: issue.message })),
  };
}
